lib/common/acl.c

```diff
@@ -676,20 +669,25 @@ pcmk__check_acl(xmlNode *xml, const char *name, enum xml_private_flags mode)
         while (parent && parent->_private) {
             xml_private_t *p = parent->_private;
             if (__xml_acl_mode_test(p->flags, mode)) {
+                g_string_free(buffer, TRUE);
                 return TRUE;
 
             } else if (is_set(p->flags, xpf_acl_deny)) {
                 crm_trace("Parent ACL denies user '%s' %s access to %s",
-                          docp->user, __xml_acl_to_text(mode), buffer);
+                          docp->user, __xml_acl_to_text(mode),
+                          (const char *) buffer->str);
                 pcmk__set_xml_flag(xml, xpf_acl_denied);
+                g_string_free(buffer, TRUE);
                 return FALSE;
             }
             parent = parent->parent;
         }
 
         crm_trace("Default ACL denies user '%s' %s access to %s",
-                  docp->user, __xml_acl_to_text(mode), buffer);
+                  docp->user, __xml_acl_to_text(mode),
+                  (const char *) buffer->str);
```

Comment:
I've only just now realized that buffer isn't used in this function for anything but trace messages. That change happened back in d307f9d (__xml_acl_check() at the time). I definitely don't want this overhead just for tracing; let's just change the messages to log tag/ref/attr instead.

We have a macro crm_str() to return "\<null\>" when a string is NULL for use in debug/trace messages, so you can use e.g. "tag %s", crm_str(tag)

That will help with the performance comparisons ...

Reply:
Ack on the overhead.

IIUC, tag/ref/attr doesn't seem applicable. AFAICT those are attributes of an XML node that represents an ACL (as in __xml_acl_create()). These log messages aren't dealing with a particular ACL. They're dealing with a CIB XML node that's being checked against ACLs.

So what do we want to print? We can print the node itself, which is not necessarily very descriptive. Getting and printing the full path requires a buffer, whether dynamic or on the stack.

Reply:
Looking back at this I'm not sure where I got tag/ref/attr from. My intent was to log the initial things that the xpath is built from, which looks like ID(xml), xml->name, and name, something like "... access to xpath built from id=%s ...". If it were user-facing, I'd want it to be friendlier, but since it's just a trace message, developers can infer the path from those if needed.

Another possibility would be to use the (relatively new) pcmk__log_else() macro to make a trace block rather than a trace message. See lib/pengine/utils.c:pe__log_node_weights() for an example. Basically you'd set the denied flag first, then do pcmk__log_else(LOG_TRACE, return FALSE), then build the buffer as before and log the message using qb_log_from_external_source() instead of crm_trace().

Reply:
> Looking back at this I'm not sure where I got tag/ref/attr from.

Presumably from __xml_acl_create(), which this commit also modifies to use GString.
(Will read over the rest of this comment later.)